apps/champions/lib/champions/battle/simulator.ex

```diff
@@ -63,187 +64,459 @@ defmodule Champions.Battle.Simulator do
+      Enum.reduce(state.units, %{}, fn {unit_id, unit}, units ->
+        if unit.health > 0 do
+          Map.put(units, unit_id, unit)
+        else
+          Logger.info("Unit #{format_unit_name(unit)} died.")
+          units
+        end
+      end)
+
+    Map.put(state, :units, new_units)
+  end
+
+  # Check if the battle has ended.
+  # Battle can end if all unit of a team are dead, or if the maximum step amount has been reached.
+  defp check_winner(state, step) do
+    winner =
+      cond do
+        Enum.empty?(state.units) -> :tie
+        Enum.all?(state.units, fn {_id, unit} -> unit.team == 2 end) -> :team_2
+        Enum.all?(state.units, fn {_id, unit} -> unit.team == 1 end) -> :team_1
+        true -> :none
+      end
+
+    case winner do
+      :none ->
+        if step == @maximum_steps do
+          Logger.info("Battle timeout.")
+          {:halt, :timeout}
+        else
+          {:cont, state}
+        end
+
+      result ->
+        Logger.info("Battle ended. Result: #{result}.")
+        {:halt, result}
+    end
+  end
+
+  # Calculate the new state of the battle after a step passes for a unit.
+  # Updates cooldowns, casts skills and reduces self-affecting modifier durations.
+  defp process_step_for_unit(unit, current_state, initial_step_state) do
     new_state =
       cond do
-        not can_attack(unit) ->
+        not can_attack(unit, initial_step_state) ->
+          Logger.info("Unit #{format_unit_name(unit)} cannot attack")
           current_state
 
         can_cast_ultimate_skill(unit) ->
-          unit
-          |> cast_skill(unit.ultimate_skill, initial_step_state, current_state)
-          |> put_in([unit.id, :energy], 0)
+          Logger.info("Unit #{format_unit_name(unit)} casting Ultimate skill")
+
+          current_state
+          |> Map.put(:skills_being_cast, [unit.ultimate_skill | current_state.skills_being_cast])
+          |> put_in([:units, unit.id, :energy], 0)
 
         can_cast_basic_skill(unit) ->
-          new_state = cast_skill(unit, unit.basic_skill, initial_step_state, current_state)
+          Logger.info("Unit #{format_unit_name(unit)} casting basic skill")
 
-          new_state
+          current_state
+          |> Map.put(:skills_being_cast, [unit.basic_skill | current_state.skills_being_cast])
           |> put_in(
-            [unit.id, :basic_skill, :remaining_cooldown],
+            [:units, unit.id, :basic_skill, :remaining_cooldown],
             unit.basic_skill.base_cooldown + 1
           )
-          |> put_in([unit.id, :energy], new_state[unit.id][:energy] + 75)
+          |> put_in([:units, unit.id, :energy], current_state.units[unit.id].energy + unit.basic_skill.energy_regen)
 
         true ->
           current_state
       end
 
-    put_in(
-      new_state,
-      [unit.id, :basic_skill, :remaining_cooldown],
-      max(new_state[unit.id].basic_skill.remaining_cooldown - 1, 0)
-    )
+    # Reduce modifier remaining timers & remove expired ones
```

Comment:
same as above, does the modifiers needs to be in the same function as the skill usage?